### PR TITLE
Add utils.py for Python models

### PR DESCRIPTION
**Pull request by Tabrizian**
Add utils.py for Python models.

### CODE REVIEW

File: src/backends/backend/python/CMakeLists.txt [hunk 1]
```diff
@@ -125,3 +129,5 @@ string(REPLACE "../../../../" "../../../../../../../../../"  PROTO_FIX_PY ${PROT
 
 install(FILES ${protocol_core_py} ${PROTO_FIX_PY} ${protocol_grpc_py} resources/startup.py
        DESTINATION lib/python/runtime)
+install(FILES ${CMAKE_CURRENT_SOURCE_DIR}/resources/utils.py
```
Comment (deadeyegoodwin) [inline src/backends/backend/python/CMakeLists.txt:132]: utils.py is too generic. Call it triton_python_backend_utils.py.

File: src/backends/backend/python/python_interpreter.cc [hunk 2]
```diff
@@ -744,9 +749,22 @@ TRITONBACKEND_ModelInstanceExecute(
   ni::ExecuteResponse execute_response;
 
   // Perform inference on the Python side
-  const auto err = instance_state->stub->Execute(
+  const auto status = instance_state->stub->Execute(
```
Comment (deadeyegoodwin) [inline src/backends/backend/python/python_interpreter.cc:752]: What if there were errors in some of the requests and so you have already sent error responses for those? How does the python code know about that? The lifecycle of requests and responses is important and is documented in the Backend section of the docs and in tritonbackend.h.
We should discuss the overall handling of this lifecycle as what is here now seems to be incorrect.

File: src/backends/backend/python/python_interpreter.cc [hunk 3]
```diff
@@ -744,9 +749,22 @@ TRITONBACKEND_ModelInstanceExecute(
...
       &context, execute_request, &execute_response);
 
+  if (!status.ok()) {
+    return TRITONSERVER_ErrorNew(
```
Comment (deadeyegoodwin) [inline src/backends/backend/python/python_interpreter.cc:756]: Lifecycle. In general once you have created and returned any response you can't return an error from this function. You must handle all response sending yourself.

File: src/backends/backend/python/resources/startup.py [hunk 4]
```diff
@@ -118,49 +117,77 @@ def __init__(self, module_path, *args, **kwargs):
...
 
     def Init(self, request, context):
+        """Init is called on TRITONBACKEND_ModelInstanceInitialize. `request`
+        object contains an args which includes a `model_config` key
+        containing the model configuration. This paramter is passed by
+        default to every ModelInstance.
+        """
         # TODO: Add error handling and returning correct status codes
         args = {x.key: x.value for x in request.args}
         self.model = self.initializer_func(args)
```
Comment (deadeyegoodwin) [inline src/backends/backend/python/resources/startup.py:127]: self.model isn't a good name. self.backend would be better.

File: src/backends/backend/python/resources/utils.py [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+# PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+# OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+import numpy as np
+
+
+class InferenceRequest:
+    """An InfrenceRequest object is used to represent Protobufs coming from
```
Comment (deadeyegoodwin) [inline src/backends/backend/python/resources/utils.py:31]: Don't say "protobuf", that is an implementation detail that is hidden from user.

Comment (deadeyegoodwin) [reply src/backends/backend/python/resources/utils.py:31]: InferenceRequest represents a request for inference for a model that executes using this backend.

File: src/backends/backend/python/resources/utils.py [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+    and simultaneously calling different methods with different threads is
+    not supported and will cause undefined behavior.
+
+    Parameters
+    ----------
+    inputs : list
+        A list of Tensor objects, each describing data for an input tensor
+        required by the model
+    request_id : str
+        Optional identifier for the request. If specified will be returned
```
Comment (deadeyegoodwin) [inline src/backends/backend/python/resources/utils.py:42]: ID assoiciated with this request, or empty string if no ID is associated with the request.

File: src/backends/backend/python/resources/utils.py [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+        in the response. Default value is an empty string which means no
+        request_id will be used.
+    correlation_id : str
```
Comment (deadeyegoodwin) [inline src/backends/backend/python/resources/utils.py:45]: Correlation ID associated with this request, or empty string if no correlation ID is associated with the request.

File: src/backends/backend/python/resources/utils.py [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+        Optional identifier for the request. If specified will be returned
+        in the response. Default value is an empty string which means no
+        request_id will be used.
+    requested_output_name : list
+        A list of strings, each describing the requested output.
```
Comment (deadeyegoodwin) [inline src/backends/backend/python/resources/utils.py:50]: The names of the output tensors that should be calculated and returned for this request.

File: src/backends/backend/python/resources/utils.py [hunk 9]
```diff
@@ -0,0 +1,160 @@
...
+        list
+            A list of Tensor objects
+        """
+        return self._output_tensors
+
+
+class Tensor:
+    """A Tensor object is used to represent inputs and output data for an
+    InferenceRequest or InferenceResponse. The object is intended to be used
+    by a single thread and simultaneously calling different methods with
```
Comment (deadeyegoodwin) [inline src/backends/backend/python/resources/utils.py:126]: These thread comments are not really correct. These are trivial classes so I don't think  you need to say anything about threading.

File: src/backends/backend/python/python_interpreter.cc (hunk 2 above)
Comment (Tabrizian) [reply src/backends/backend/python/python_interpreter.cc:753]: I release all the requests if an error happens during inference.

File: src/backends/backend/python/resources/triton_python_backend_utils.py [hunk 10]
```diff
@@ -0,0 +1,155 @@
...
+        -------
+        list
+            A list of Tensor objects
+        """
+        return self._output_tensors
+
+
+class Tensor:
+    """A Tensor object is used to represent inputs and output data for an
+    InferenceRequest or InferenceResponse. The object is intended to be used
```
Comment (deadeyegoodwin) [inline src/backends/backend/python/resources/triton_python_backend_utils.py:120]: Remove the thread part of the comment

**Pull request merged**